Add support for collapsing notebook sections via the ToC

kgryte opened the pull request:
![collapsible_headers](https://user-images.githubusercontent.com/2643044/66435384-8e888f00-e9d9-11e9-9b33-5d259cf7d387.gif)

This PR adds support for collapsing notebook sections via the ToC. This PR resolves https://github.com/jupyterlab/jupyterlab-toc/issues/12 and can partially address the desire for collapsing all sections as requested in https://github.com/jupyterlab/jupyterlab-toc/issues/65.

## Notes

-   Collapsing sections only works one way: collapsing a ToC element collapses a notebook section. This PR does **not** add support for collapsing a ToC element by manually collapsing a notebook section. The reason this PR does not provide bi-directional support is as discussed in https://github.com/jupyterlab/jupyterlab/issues/2275.

    Notably, while manually collapsing a notebook section may be complementary to the behavior introduced by this PR, such behavior should not require this particular extension (i.e., as a user, I may want to collapse notebook sections but not want a ToC). Accordingly, modifying the notebook to include UI elements for manually collapsing notebook sections has been avoided with preference for this being either built-in to JupyterLab or implemented as a separate extension.

    As noted in the source, we can imagine a scenario in which the underlying implementation may change if, and when, such section collapsing behavior is implemented elsewhere. However, for the time being, we manually set the `hidden` state of heading defined notebook sections (i.e., as defined according to the heading hierarchy).

-    In contrast to the original implementation, all sections, including sections without "children", are collapsible in the ToC and in the notebook.

-   While the notebook file format does include a [`collapsed`](https://nbformat.readthedocs.io/en/latest/format_description.html?highlight=collapsed#metadata) meta data field, unfortunately, this field cannot be used in this particular use case, as this field, as discussed and clarified [elsewhere](https://github.com/jupyter/nbformat/issues/137), only applies to an individual cell's output container and does not extend to the more abstract grouping imposed by a heading hierarchy.

kgryte commented:
Potential reviewers: @ian-r-rose

kgryte commented:
@saulshanabrook Yeah, handling reordering would seem fodder for a separate PR.

Note: #102 should be merged before this PR, as this PR incorporates changes from that pull request (e.g., the inclusion of the demo notebook).

ellisonbg commented:
@kgryte fantastic - thanks for diving into this! This has been on my todo list since 2011 ;-)

I will start to review/test, but an initial set of questions:

* Where are we storing the collapsed state?
* As we move to real time collaboration, should that state be synchronized across users?

ellisonbg commented:
Here is the Binder link: https://mybinder.org/v2/gh/kgryte/jupyterlab-toc/collapsible-headers?urlpath=lab%2Ftree%2Fnotebooks%2Fdemo.ipynb

ellisonbg commented:
In testing I found that if I move a markdown heading cell when the section is collapsed, only the heading cell gets moved. When you un-collapse the heading you just moved, the children cells show again, but in their original location. I think the entire section should be moved when a collapsed heading cell is moved. It looks like cut/copy/paste are disabled for collapsed sections, at a minimum, we should disable moving cells (through drag/drop) until the full behavior is fixed.

ellisonbg commented:
Other minor point, the visual design of the TOC attemps to be very compact - I think the twisty icons are proportionally larger than the other elements - so maybe try going down in size a tad?

ellisonbg commented:
But overall, the experience is pretty spectacular. We have needed this for a really long time and I am super happy to see it move forward!

kgryte commented:
@ellisonbg Responses:

> 1. Where are we storing the collapsed state?

I am storing the cell collapsed state in the cell `metadata` field. This is where the ToC header collapsed state is also stored, so I simply followed precedent.

> 2. As we move to real time collaboration, should that state be synchronized across users?

The answer to this question seems out of scope for this PR and certainly beyond my knowledge of how RTC is being implemented.

> 3. Drag and drop.

Indeed, the only element which moves is the heading; however, not clear to me that moving the entire "section" makes sense either as the sole mode of drag and drop. As a user, I can imagine simply wanting to move a heading element and not the entire section. And I can imagine wanting to move an entire "section" as determined by the heading.

With regard to this PR and ToC, in general, the ToC should simply reflect the notebook layout. And, in fact, this is what it does. If we want to allow moving "sections", this would seem fodder for a separate extension which can "group" cells into sections (probably in a manner similar to how the ToC extension determines groupings: by analyzing the heading hierarchy) and treat them as units which can be reordered as a whole. That functionality would seem useful on its own and should not be dependent on a ToC and, instead, complementary.

Provided such an extension was installed alongside the ToC extension, when a notebook "section" was moved, the ToC would, as it stands now, automatically update, as it simply reflects the current notebook cell order.

> 4. Twisty icon size.

Adjusting the icon sizing can be addressed in a subsequent PR, as changing the design elements (styling) seems beyond the scope of the current PR. When I have some cycles to spare, I can experiment locally and submit another PR for your review.

kgryte commented:
@ellisonbg Also, thanks for the feedback and for trying it out!

ian-r-rose commented:
Thanks for this @kgryte! The behavior looks pretty good, issues around reordering notwithstanding. A few comments on the implementation:

>     * Collapsing sections only works one way: collapsing a ToC element collapses a notebook section. This PR does **not** add support for collapsing a ToC element by manually collapsing a notebook section. The reason this PR does not provide bi-directional support is as discussed in [jupyterlab/jupyterlab#2275](https://github.com/jupyterlab/jupyterlab/issues/2275).

Can you elaborate a bit on why we don't want to include bi-directional support here? More on this below...

>       Notably, while manually collapsing a notebook section may be complementary to the behavior introduced by this PR, such behavior should not require this particular extension (i.e., as a user, I may want to collapse notebook sections but not want a ToC). Accordingly, modifying the notebook to include UI elements for manually collapsing notebook sections has been avoided with preference for this being either built-in to JupyterLab or implemented as a separate extension.

I agree that we likely don't want to modify the notebook UI if we don't have to. We already do that a bit with the auto-numbering, and its implementation makes me a bit uncomfortable.

>       As noted in the source, we can imagine a scenario in which the underlying implementation may change if, and when, such section collapsing behavior is implemented elsewhere. However, for the time being, we manually set the `hidden` state of heading defined notebook sections (i.e., as defined according to the heading hierarchy).

I'm not sure I agree with this choice -- I think there should be some visual indication of when there is a collapsed cell. Furthermore, we *already* have a collapsing UI for cells. I was hoping to use that logic, and its associated APIs, rather than raw DOM manipulation. That is to say, I think we should try, whenever possible, to work within the existing UX and APIs.

>     * While the notebook file format does include a [`collapsed`](https://nbformat.readthedocs.io/en/latest/format_description.html?highlight=collapsed#metadata) meta data field, unfortunately, this field cannot be used in this particular use case, as this field, as discussed and clarified [elsewhere](https://github.com/jupyter/nbformat/issues/137), only applies to an individual cell's output container and does not extend to the more abstract grouping imposed by a heading hierarchy.

I don't agree with this -- we also have an `inputs_hidden` field, so we can, in fact, use the existing metadata fields to determine whether an entire cell is collapsed. This same logic is used to determine how to render the existing collapsers for the cell widgets. I've been wanting to remove the TOC cell metadata for a while -- we already have redundant official metadata fields for collapsing, and adding yet another metadata field would further confuse things.

Alternatively, we can put the TOC view state in the `IStateDB` (another thing that's been on my mind for a while), and not mess with the cell metadata at all. This would speak to the RTC question that @ellisonbg brought up. We have talked a lot about what "view-related" stuff should be synchronized, and mostly don't have firm answers. But if we do use cell metadata, then the collapsed state would likely be synchronized.

I wasn't able to reproduce the issue that @saulshanabrook saw -- when I reordered cells the ToC seemed to update okay. I did find the behavior of moving collapsed cells to be confusing -- it didn't seem to update the collapsed state of the children properly if the parent header was moved. I think it's fine to leave that for a follow-up, however.

ian-r-rose commented on src/generators/notebookgenerator/itemrenderer.tsx:
```diff
@@ -13,8 +15,161 @@ import { sanitizerOptions } from '../shared';
+  if (cell.constructor.name !== 'MarkdownCell') {
+    return -1;
+  }
+  const lines = cell.model.value.text.split('\n');
```
This stuff is fiddly, and there are a lot of edge cases. I think we should refactor to use the same logic that is used to generate cell headings, or otherwise reuse the `INotebookHeading` for a cell.

kgryte replied:
I followed the same general [logic](https://github.com/jupyterlab/jupyterlab-toc/blob/2a3291ae31ce6f63b470c6865e43f39e23e12b61/src/generators/notebookgenerator/index.ts#L411) as used to generate Markdown headings, so simply followed precedent.

ian-r-rose commented on src/generators/notebookgenerator/itemrenderer.tsx:
```diff
@@ -13,8 +15,161 @@ import { sanitizerOptions } from '../shared';
+  if (level === -1) {
+    return;
+  }
+  const widgets = tracker.currentWidget.content.widgets;
```
`tracker.currentWidget` can be null, we may want to double check that it exists.

kgryte replied:
Done.

ian-r-rose commented on src/generators/notebookgenerator/itemrenderer.tsx:
```diff
@@ -13,8 +15,161 @@ import { sanitizerOptions } from '../shared';
+ * @param tracker - notebook tracker
+ * @param cell - notebook cell
+ */
+function uncollapseCell(tracker: any, cell: Cell): void {
```
This function is very similar to the above. Maybe they can be generalized to something like `setCollapseState` with a boolean parameter to collapse/uncollapse?

kgryte replied:
Done.

ian-r-rose commented on src/generators/notebookgenerator/itemrenderer.tsx:
```diff
@@ -13,8 +15,161 @@ import { sanitizerOptions } from '../shared';
+    w.setHidden(false);
+  }
+  // Set a meta-data flag to indicate that we've un-collapsed notebook sections:
+  cell.model.metadata.set('toc-nb-collapsed', false);
```
I'm still not convinced about the choice to use metadata here (which was not your choice, @kgryte), but as long as we are using it -- I think we should remove the metadata here rather than setting to `false`.

kgryte replied:
Addressed. Deleting a field may have a negative effect on performance, but likely to not be rate-limiting given everything else which is happening when un-collapsing a section.

ian-r-rose commented on src/generators/notebookgenerator/itemrenderer.tsx:
```diff
@@ -23,8 +178,16 @@ export function notebookItemRenderer(
         'toc-hr-collapsed'
       ) as boolean;
-      collapsed = collapsed != undefined ? collapsed : false;
+      collapsed = collapsed !== undefined ? collapsed : false;
```
Ooh, I'd really like to turn on `strictNullChecks` for this project.

kgryte replied:
Yeah, agreed. However, this sort of thing (i.e., loose equality tests) is present throughout the codebase. Such a change should be part of a separate PR aimed at improving code quality and will likely require a bit of work. I had initially tried to clean up a couple other files, but gave up once I started encountering compiler errors and came to see it to as a time sink.

ian-r-rose commented on src/generators/notebookgenerator/itemrenderer.tsx:
```diff
@@ -23,8 +178,16 @@ export function notebookItemRenderer(
-      collapsed = collapsed != undefined ? collapsed : false;
+      collapsed = collapsed !== undefined ? collapsed : false;
       cellRef!.model.metadata.set('toc-hr-collapsed', !collapsed);
+      if (cellRef && tracker) {
```
Why the `tracker` check here? Is it allowed to be `null`?

kgryte replied:
Removed.

ian-r-rose commented on src/generators/notebookgenerator/itemrenderer.tsx:
```diff
@@ -23,8 +178,16 @@ export function notebookItemRenderer(
+      if (cellRef && tracker) {
+        // NOTE: we can imagine a future in which this extension combines with a collapsible-header/ings extension such that we can programmatically close notebook "sections". In the meantime, we need to resort to manually "collapsing" sections...
+        if (collapsed) {
+          uncollapseCell(tracker as any, cellRef);
```
Let's figure out the typings so we don't need this cast.

kgryte replied:
Done.

ian-r-rose commented on src/generators/notebookgenerator/itemrenderer.tsx:
```diff
@@ -13,8 +15,161 @@ import { sanitizerOptions } from '../shared';
+ * @param tracker - notebook tracker
+ * @param cell - notebook cell
+ */
+function collapseCell(tracker: any, cell: Cell): void {
```
Why is `tracker` `any`? Isn't it of type `INotebookTracker`?

ian-r-rose replied:
I think it may be better to pass in the `Notebook` widget, rather than the `INotebookTracker` here. This makes the universe that this function needs to know about smaller.

kgryte replied:
Addressed.

ian-r-rose commented:
@ellisonbg I agree that the twist looks a bit big. At one point we also had some indentation for subheadings, and that appears to have gone away. I think we should restore it, but I don't really remember why it went away to begin with, so perhaps I am missing something.

kgryte commented on src/generators/notebookgenerator/itemrenderer.tsx:
````diff
@@ -13,8 +15,161 @@ import { sanitizerOptions } from '../shared';
+ *     -   HTML heading:
+ *
+ *         ```
+ *         <h3>Foo</h3>
````
Was unaware of this use case. Will explore.

kgryte replied:
@ian-r-rose Working on this now. Supporting cell outputs significantly increases complexity and introduces a bit of a rabbit hole, especially given that the current ToC support for cell outputs has buggy behavior.

kgryte commented:
@ian-r-rose For context, this is what happens when you use the existing [nbformat](https://nbformat.readthedocs.io/en/latest/format_description.html?highlight=collapsed#cell-metadata) meta data fields (`collapsed`, `source_hidden`, and `output_hidden`):

![image](https://user-images.githubusercontent.com/2643044/66795483-9c4a8280-eeb9-11e9-9012-e8677358b328.png)

I, too, would have preferred to leverage existing UX and APIs; however, because of how the Jupyter notebook renderer interprets those fields and subsequently behaves, I did not feel that setting them was appropriate. Unfortunately, setting them entails certain side effects, which did not seem appropriate for the desired intent.

Accordingly, I leveraged the `setHidden` widget API which seemed more appropriate and more in accordance with user expectation.

kgryte commented:
@ian-r-rose I've added partial support for expanding/collapsing sections generated via code cells, and updated the demo notebook to show both various edge cases (and bugs) and a recommended approach (namely, hardcoding section headings).

Providing full parity for cell outputs will require a detailed rethink in terms of how and what we track to ensure proper collapsing behavior. As the user story for cell outputs is not the dominant use case, I am inclined to punt on achieving full parity until a future time.

Some of the cell output bugs are most easily conveyed via demo. If you want to hop on a video call, I can walk you through the edge cases.

Otherwise, I feel like this PR satisfies most user stories, and, provided a user follows the recommended approach when generating Markdown as discussed in the demo notebook, should work as expected.

ian-r-rose commented:
I could meet at around 6PM PST, if you are still available then @kgryte

kgryte commented:
@ian-r-rose Works for me!

kgryte commented:
@ian-r-rose Okay. Should be ready for a second review. Per discussion, I went ahead and refactored/"cleaned" the codebase. Accordingly, **forewarning**, nearly every implementation file was touched. In which case, you may want to pull down my fork locally, check out the PR branch, and review from your editor.

A couple notes regarding the refactoring:

1.  I reorganized the codebase to better adhere to JavaScript/Node.js best practices. Notably, for the most part, small files, single-file single-export, hoisted function definitions, and a limited publicly exposed API surface.
2.  Made parameter order consistent across similar APIs (e.g., ToC generators).
3.  Added source code documentation throughout the codebase.
4.  Renamed methods and variables to be less verbose, while still, IMO, maintaining readability.
5.  Where obvious, addressed approaches having suboptimal performance (e.g., favoring loops over functional APIs which incur stack overhead).

Hopefully, the above addresses some of your concerns regarding future maintainability of the extension and better positions the repository for further development going forward.

aquirdTurtle commented:
Are guys sure you want to add collapsing notebook headers to the TOC extension? The behavior seems a bit orthogonal to the idea of a table of contents, and an extension for collapsible headers already exists here - https://github.com/aquirdTurtle/Collapsible_Headings.

kgryte commented:
@aquirdTurtle As stated in the OP, the behavior implemented here is uni-directional. In short, a common ToC request (and as provided in the older `toc2` extension) is the ability to collapse notebook headings **via** the ToC panel. This is what this PR does.

This PR does **not** implement functionality for collapsing headings directly in a notebook, which what your extension does.

mlucool commented:
> Are you all sure you want to add collapsing notebook headers to the TOC extension?

While I agree it is somewhat tangent, I think the ToC is a great place for it in the end. First, it moves the idea of HTML header levels into an organized structure, then it gives a nice UI to see what you will collapse when collapsing. Really, the ToC seem like a nice place to put anything that needs to assume a hierarchy from markdown.

Having it inline in the notebook is a subpar UX for a couple of reasons. Firstly, you can't easily tell what will be collapsed. It's hard to tell an h3 from an h4 visually but easy to see in tree form. Secondly, the collapse inline add complexity to the notebook itself by having two elements which collapse different things next to each other (the other being the blue collapse cell). I found that somewhat distracting.

While I do think you have a very interesting plugin, I personally prefer the approach proposed here. Do you disagree with my reasoning @aquirdTurtle?

marius311 commented:
My 2cents as a user, ideally one can collapse sections either from the notebook or from the TOC, and its an option left to the user whether collapsing a section from the notebook gets reflected in the TOC and vice-versa (I personally would never want to collapse the TOC, but I can definitely see either preference). This was the behavior on Jupyter notebook with the old toc2 and collapsible_headings extensions, which I found very useful and would hope eventually gets reproduced for Jupyterlab.

aquirdTurtle commented:
@mlucool I think in the end it mostly boils down to user preference. 

>  It's hard to tell an h3 from an h4 visually but easy to see in tree form.

I could argue that this is rather an issue with the header rendering rather than a natural advantage of the TOC-style.  For some context, one of the main competitors to Jupyter (& python in general) within the physics community is the Wolfram Mathematica notebook (largely popular due to its stellar symbolic math support), which has a collapsible headings style closer to the in-notebook style, so that's where I'm coming from. 

Anyways I'm not trying to stop this nice feature from being included in the extension or anything (as if I could!) I just thought I'd ask a provocative question :)

ian-r-rose commented on src/toc.tsx:
```diff
@@ -1,47 +1,82 @@
+/**
+ * Interface describing table of contents widget options.
+ *
+ * @private
```
Why have you marked this as private? It is part of the public interface of the `TableOfContents` widget.

kgryte replied:
You're right. Removed.

ian-r-rose commented on src/toc.tsx:
```diff
@@ -1,47 +1,82 @@
- * Timeout for throttling TOC rendering.
+ * Interface describing the current widget.
+ *
+ * @private
```
Why have you marked this as private? It is also part of the public interface. I think it would be best to have an `ITableOfContents` namespace for the this and the above `IOptions`. This is mostly equivalent to static class attributes, but more in keeping with previous code style,

kgryte replied:
Removed. Re: namespace. Not of the opinion that this is necessary, as, e.g., in an editor like VSCode, the main intent is to describe the "shape" of data, rather than defining a type to which one casts. Removing the `@private` annotations here makes sense from the perspective of using `typedoc` and allowing readers to inspect input and output data shapes. So went with just removing the `@private` annotations.

ian-r-rose replied:
I agree that it is not necessary, it is more of an issue of following a consistent style and set of design patterns which is used in JupyterLab. See, for instance, the same pattern [here](https://github.com/jupyterlab/jupyterlab/blob/f2e0cde0e7c960dc82fd9b010fcd3dbd9e9b43d0/packages/docmanager/src/manager.ts#L593-L639).

ian-r-rose replied:
I'm actually a little surprised this compiles: these interfaces should be public and usable by other extensions, but are not exported by this module.

kgryte replied:
There may be some confusion here as to what [`@private`](https://jsdoc.app/tags-private.html) does. This annotation originates from JSDoc and does not affect "usability" in any sense, which is in contrast to TypeScript's notion of [`private`](https://www.typescriptlang.org/docs/handbook/classes.html#understanding-private) class members. The code compiled because it is not necessary to export them or have them be annotated as public, as they simply describe the shape of data and are compiled away. The original reason I added the `@private` annotation was in (mistakingly) thinking that they would not be needed displayed in generated TypeDoc documentation. Removing the annotations allows them to be.

Regardless, if the desired convention is to export them as part of a namespace, the above is a moot point.

ian-r-rose commented on src/toc.tsx:
```diff
@@ -146,164 +175,35 @@ export class TableOfContents extends Widget {
-  protected onAfterShow(msg: Message): void {
+  protected onUpdateRequest(msg: Message): void {
+    // Don't bother if the ToC is not visible...
+    /* if (!this.isVisible) {
```
~~Is this commented out for a reason?~~ Edit: I see it was that way already. Feel free to remove, or I can clean it up later if not.

kgryte replied:
Removed.

ian-r-rose commented on src/toc.tsx:
```diff
@@ -146,164 +175,35 @@ export class TableOfContents extends Widget {
+    /* if (!this.isVisible) {
+      return;
+    } */
     this.update();
```
~~This is circular. A call to `update` triggers an `onUpdateRequest` call in the next stack frame, so this will essentially animate the ToC, doing a lot of extra work.~~

Edit: nevermind, see above.

ian-r-rose commented on src/toc.tsx:
```diff
@@ -55,45 +90,34 @@ export class TableOfContents extends Widget {
-  }
-
-  updateTOC() {
+  update() {
```
This overrides an existing `update()` method, which we don't want to hide. Let's revert the name change, or put this logic into the `onUpdateRequest` function.

kgryte replied:
Moved logic to the `onUpdateRequest` function. Thanks! Did not realize I clobbered an existing method.

ian-r-rose commented on src/toc_tree.tsx:
```diff
@@ -0,0 +1,88 @@
+  /**
+   * Toolbar.
+   */
+  toolbar: any;
```
Memo to myself: we should tighten up the typings here.

kgryte replied:
Yeah, was not sure the original intent, so left as is.

ian-r-rose commented on src/utils/parse_heading.ts:
```diff
@@ -0,0 +1,115 @@
+  /**
+   * Heading type.
+   */
+  type: 'html' | 'markdown' | 'markdown-alt';
```
Interesting. What is the purpose of differentiating based on these different styles of markdown heading?

kgryte replied:
Because this logic is moved to its own utility function, needs to be generic enough to handle, e.g., the case where you parse a string for a heading, but then, based on the heading type, want to do further manipulation. That manipulation may depend on what type of heading was recovered. For alternative Markdown formatting, that might include removing the second line and replacing the first with `#####` style headings, etc. So, in short, as I plucked this logic from other contexts, I opted for retaining the "type" of the heading recovered in the spirit of a general utility.

ian-r-rose commented on src/generators/notebook/set_collapsed_state.ts:
```diff
@@ -0,0 +1,157 @@
+      for (let k = 0; k < dtypes.length; k++) {
+        const t = dtypes[k];
+        if (isMarkdown(t) || isDOM(t) || t === 'text/plain') {
+          // FIXME: apparent Jupyter bug where additional Markdown displays have a `text/plain` MIME type
```
Ooh, interesting. Is there an issue for this? Can you provide a reproducible example?

kgryte replied:
I did not find a related issue, but I cannot remember how hard I looked either. Basically, I discovered this in the course of generating multiple Markdown displays and logging `t`, noticing that the MIME type was not the expected Markdown MIME type.

ian-r-rose commented on src/generators/notebook/set_collapsed_state.ts:
```diff
@@ -0,0 +1,157 @@
+      continue;
+    }
+    // Code cells are more involved, as we need to analyze the outputs to check for generated Markdown/HTML...
+    const c = w as CodeCell;
```
Yikes, I didn't know this was going to be so involved! Thanks for doing all the hard work here!

kgryte replied:
Yeah, ;) code cell output is hard, involved, and entails considerable complexity, as discussed elsewhere. This provides enough support, IMO, for the majority of output cell use cases.

ian-r-rose commented on src/generators/notebook/set_collapsed_state.ts:
```diff
@@ -0,0 +1,157 @@
+      ow.setHidden(state);
+    }
+    // Collapse/expand a sub-cell's input area by setting the its `hidden` state (NOTE: a collapsed input area will still appear collapsed!!! Once a user clicks on the collapsed cell ellipses will the input area display (or, rather, not display!) as being fully hidden. If we want an input area to always be present, e.g., to always allow a user the ability to edit an expanded section, whether or n ...
+    // w.inputArea.setHidden(state);
```
Is this intended to be commented out?

kgryte replied:
Yes, I added a comment. In short, was not absolutely clear to me the desired behavior. Ended up deciding to avoid collapsing/expanding a sub-cell's input area because I preferred the rendered output better. However, I wanted to leave in the code comments so that a future dev reading the code might more easily discover the option that rendering could be different or even preferred.

ian-r-rose commented on src/utils/generate_numbering.ts:
```diff
@@ -0,0 +1,64 @@
+const MAX_HEADING_LEVEL = 6;
+
+/**
+ * Updates numbering dictionary levels.
```
Can we say in what way this updates it?

kgryte replied:
TBH I never really understood the intent behind the original logic or why mutation was necessary. I think the main rationale for this being in a separate function in the original codebase was to circumvent compiler errors related to setting levels to non-numeric values. But this is only a guess.

Nor did I trace how this mutation might be depended on throughout the application.

So, I punted on this, opting for a (less than optimal) generic description, which I think is fine, for this particular case, as it is an internal helper function which simply encapsulates some logic.

ian-r-rose commented on notebooks/demo.ipynb:
```diff
@@ -455,6 +449,515 @@
+    "## Notes\n",
+    "\n",
+    "-   Markdown cells containing multiple \"sections\" only show as the first section in the ToC. Accordingly, users should add a **new** Markdown cell for **each** section.\n",
+    "-   The ToC extension **does** detect multiple \"sections\" when those sections are generated by code cells. However, users are likely to run into buggy ToC behavior when doing so. Users are strongly encouraged to **avoid** using code cells to generate section headings.\n",
```
I'd rather not advertise vague buggy behavior here. If it is indeed a bug, let's briefly describe the current behavior, say why it's a problem, then open an issue on the repo.

kgryte replied:
I updated the note to include an example. I am not convinced it's a issue worthy bug rather than a user "error" in doing things which are generally not recommended, but are allowed within Jupyter.

kgryte commented on src/extension.ts:
```diff
@@ -63,67 +46,99 @@ function activateTOC(
-  // Create the ToC registry.
-  const registry = new TableOfContentsRegistry();
+  // Create the ToC registry:
+  const registry = new Registry();
```
Renamed. Was unaware of this convention.

kgryte commented on src/generators/index.ts:
```diff
@@ -1,6 +1,7 @@
-export * from './markdowndocgenerator';
-export * from './latexgenerator';
-export * from './notebookgenerator';
+// Note: keep in alphabetical order...
```
Thinking toward a potential future in which there are many (?) ToC generators. In which case, having in alphabetical order allows for faster scanning when looking for, in this case, a particularly export. This may be me being slightly pedantic, but my tendency is to prefer imposing order over disorder.

kgryte commented:
@ian-r-rose Thanks for the review!

Regarding the three issues:

1. This should now be resolved. I updated the [`render`](https://github.com/kgryte/jupyterlab-toc/blob/7528c754d94e1ce083ec54c139da74958a8e2758/src/generators/notebook/render.tsx#L56) method to set the notebook cell collapsed state at the same time as the ToC twist buttons are rendered.

2. TBH I had not tested behavior with cell tags, as I was never able to figure out how to add cell tags to begin with. From what I can [see](https://github.com/kgryte/jupyterlab-toc/blob/a3a0b1759324754bf7243600273b05e23c12e9c9/src/generators/notebook/tagstool/index.tsx#L74), we should already be triggering a re-render of the table of contents, but maybe I am not looking in the right place. Maybe this will be resolved now that I moved the `update` logic to `onUpdateRequest`, no longer clobbering the inherited method? Not sure.

3. Not sure what the behavior you are seeing is. Have you tried reproducing it in the current `master` branch? Tis possible that my refactoring introduced a regression, but I am not sure where. I am also not sure what you mean by Markdown "preview".

Otherwise, I believe I have resolved the PR feedback and pushed all the updated code, so should be ready for another round of review, unless items 2 & 3 need to be resolved beforehand. LMK!

ian-r-rose commented:
I've come across one more lifecycle issue: when a TOC generator does not provide a toolbar, the toolbar from the previously focused notebook sticks around. To reproduce:

1. Open a notebook.
2. Open a LaTeX document. It will inherit the notebook toolbar.

This does not appear to be a problem introduced by this PR, so I'm happy to look into it next week.

ian-r-rose commented:
Okay, here is my plan:
1. We merge this!
1. I publish an RC version of this extension so people can try it out.
1. Next week I try to track down the remaining lifecycle issues.
1. We publish a final version.

ian-r-rose commented:
Thanks for all the hard work on this @kgryte!

ian-r-rose commented:
Published as 2.0.0-rc.0

jph00 commented:
> Notably, while manually collapsing a notebook section may be complementary to the behavior introduced by this PR, such behavior should not require this particular extension (i.e., as a user, I may want to collapse notebook sections but not want a ToC). Accordingly, modifying the notebook to include UI elements for manually collapsing notebook sections has been avoided with preference for this being either built-in to JupyterLab or implemented as a separate extension.

@kgryte do you know if anyone is working on this? It's one of the things that's kept us from using JupyterLab for fastai and nbdev.

PS: thanks a lot for this great PR!

kgryte commented:
@jph00 To my knowledge, not in a general sense for notebook Markdown cells containing multiple headings. But, with some restrictions, there is an [extension](https://github.com/aquirdTurtle/Collapsible_Headings) which adds UI elements to manually collapsing headings (as mentioned earlier in this thread) provided the document model is simple (i.e., one heading per Markdown cell).

jph00 commented:
Oh that's great! Sorry I missed that earlier link.
